File: app/src/main/java/sk/styk/martin/apkanalyzer/business/service/statistics/LocalApplicationStatisticDataService.java
package sk.styk.martin.apkanalyzer.business.service.statistics;

import android.content.pm.ApplicationInfo;
import android.content.pm.PackageInfo;
import android.content.pm.PackageManager;

import sk.styk.martin.apkanalyzer.business.service.CertificateService;
import sk.styk.martin.apkanalyzer.business.service.FileDataService;
import sk.styk.martin.apkanalyzer.business.service.GeneralDataService;
import sk.styk.martin.apkanalyzer.business.service.ResourceService;
import sk.styk.martin.apkanalyzer.model.detail.FileData;
import sk.styk.martin.apkanalyzer.model.detail.ResourceData;
import sk.styk.martin.apkanalyzer.model.statistics.LocalStatisticsAppData;

/**
 * Created by Martin Styk on 28.07.2017.
 */
public class LocalApplicationStatisticDataService {

    private PackageManager packageManager;

    private GeneralDataService generalDataService;
    private CertificateService certificateService;
    private FileDataService fileService;
    private ResourceService resourceService;

    public LocalApplicationStatisticDataService(PackageManager packageManager) {
        this.packageManager = packageManager;
        this.generalDataService = new GeneralDataService(packageManager);
        this.certificateService = new CertificateService(packageManager);
        this.fileService = new FileDataService(packageManager);
        this.resourceService = new ResourceService(packageManager);
    }

    public LocalStatisticsAppData get(String packageName) {
        PackageInfo packageInfo = null;
        try {
            packageInfo = packageManager.getPackageInfo(packageName,
                    PackageManager.GET_SIGNATURES |
                            PackageManager.GET_ACTIVITIES |
                            PackageManager.GET_SERVICES |
                            PackageManager.GET_PROVIDERS |
                            PackageManager.GET_RECEIVERS |
                            PackageManager.GET_PERMISSIONS);
        } catch (PackageManager.NameNotFoundException e) {
            e.printStackTrace();
            return null;
        }
        ApplicationInfo applicationInfo = packageInfo.applicationInfo;

        if (packageInfo == null || applicationInfo == null)
            return null;

        LocalStatisticsAppData data = new LocalStatisticsAppData();
        data.setSystemApp((applicationInfo.flags & ApplicationInfo.FLAG_SYSTEM) != 0);
        data.setInstallLocation(packageInfo.installLocation);
        data.setTargetSdk(applicationInfo.targetSdkVersion);
        data.setMinSdk(generalDataService.getMinSdkVersion(packageName));
        data.setApkSize(generalDataService.getApkSize(applicationInfo.sourceDir));

        data.setSignAlgorithm(certificateService.getSignAlgorithm(packageInfo));

        data.setActivites(packageInfo.activities.length);
        data.setServices(packageInfo.services.length);
        data.setProviders(packageInfo.providers.length);
        data.setReceivers(packageInfo.receivers.length);

        data.setDefinedPermissions(packageInfo.permissions.length);
        data.setUsedPermissions(packageInfo.requestedPermissions.length);

        FileData fileData = fileService.get(packageInfo);
        data.setFiles(fileData.getAllHashes().size());

        ResourceData resourceData = resourceService.get(fileData);
        data.setDrawables(resourceData.getDrawables());
        data.setDifferentDrawables(resourceData.getDifferentDrawables());
        data.setLayouts(resourceData.getLayouts());
        data.setDifferentLayouts(resourceData.getDifferentLayouts());

        return data;
    }

}

File: app/src/main/java/sk/styk/martin/apkanalyzer/business/analysis/logic/launcher/LocalApplicationStatisticDataService.kt
package sk.styk.martin.apkanalyzer.business.analysis.logic.launcher

import android.content.pm.ApplicationInfo
import android.content.pm.PackageManager
import sk.styk.martin.apkanalyzer.business.analysis.logic.AndroidManifestService
import sk.styk.martin.apkanalyzer.business.analysis.logic.CertificateService
import sk.styk.martin.apkanalyzer.business.analysis.logic.FileDataService
import sk.styk.martin.apkanalyzer.business.analysis.logic.GeneralDataService
import sk.styk.martin.apkanalyzer.business.analysis.logic.ResourceService
import sk.styk.martin.apkanalyzer.model.statistics.LocalStatisticsAppData

/**
 * @author Martin Styk
 * @version 28.07.2017.
 */
class LocalApplicationStatisticDataService(private val packageManager: PackageManager) {

    private val analysisFlags = PackageManager.GET_SIGNATURES or
            PackageManager.GET_ACTIVITIES or
            PackageManager.GET_SERVICES or
            PackageManager.GET_PROVIDERS or
            PackageManager.GET_RECEIVERS or
            PackageManager.GET_PERMISSIONS


    private val generalDataService: GeneralDataService = GeneralDataService()
    private val certificateService: CertificateService = CertificateService()
    private val fileService: FileDataService = FileDataService()
    private val resourceService: ResourceService = ResourceService()

    fun get(packageName: String): LocalStatisticsAppData? {

        val packageInfo = try {
            packageManager.getPackageInfo(packageName, analysisFlags)
        } catch (e: Exception) {
            e.printStackTrace()
            return null
        }

        val applicationInfo = packageInfo.applicationInfo ?: return null

        val isSystemApp = (applicationInfo.flags and ApplicationInfo.FLAG_SYSTEM) != 0
        val fileData = fileService.get(packageInfo)
        val resourceData = resourceService.get(fileData)

        return LocalStatisticsAppData(
                packageName = packageName,
                isSystemApp = isSystemApp,
                installLocation = packageInfo.installLocation,
                targetSdk = applicationInfo.targetSdkVersion,
                minSdk = AndroidManifestService.getMinSdkVersion(applicationInfo, packageManager) ?: 0,
                apkSize = if (applicationInfo.sourceDir != null) generalDataService.computeApkSize(applicationInfo.sourceDir) else 0,
                appSource = GeneralDataService.getAppSource(packageManager, packageName, isSystemApp),
                signAlgorithm = certificateService.getSignAlgorithm(packageInfo) ?: "Unknown",
                activities = packageInfo.activities?.size ?: 0,
                services = packageInfo.services?.size ?: 0,
                providers = packageInfo.providers?.size ?: 0,
                receivers = packageInfo.receivers?.size ?: 0,
                definedPermissions = packageInfo.permissions?.size ?: 0,
                usedPermissions = packageInfo.requestedPermissions?.size ?: 0,
                files = fileData.totalFiles,
                drawables = resourceData.drawables,
                differentDrawables = resourceData.differentDrawables,
                layouts = resourceData.layouts,
                differentLayouts = resourceData.differentLayouts
        )
    }

}
